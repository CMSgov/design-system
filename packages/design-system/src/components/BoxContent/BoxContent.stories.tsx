import type { Meta, StoryObj } from '@storybook/react';
import BoxContent from './BoxContent';

const meta: Meta = {
  component: BoxContent,
  argTypes: {
    children: { control: 'text' },
    bordered: { control: 'boolean', defaultValue: true },
    heading: { control: 'text' },
    headingLevel: { control: 'text' },
  },
  parameters: {
    docs: {
      underlyingHtmlElements: ['aside'],
    },
  },
};

export default meta;
type Story = StoryObj<typeof BoxContent>;

const BoxContentTemplate: Story = {
  render: ({ ...args }) => {
    return <BoxContent {...args}>{args.children}</BoxContent>;
  },
};

export const Default = {
  ...BoxContentTemplate,
  args: {
    heading: 'The Inflation Reduction Act',
    children:
      "The Inflation Reduction Act keeps these savings and lower costs through 2025. If you qualify for savings, you'll find out the lower costs when you shop for plans.",
    bordered: true,
  },
};
