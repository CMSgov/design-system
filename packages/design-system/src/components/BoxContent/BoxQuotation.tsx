import type * as React from 'react';
import { FunctionComponent } from 'react';
import { RequireAtLeastOne } from '../utilities/requireAtLeastOne';
import classNames from 'classnames';

interface MinimumBoxQuotationProps {
  /**
   * Provide an author for the quote
   */
  author?: string;
  /**
   * Content to be displayed within the Box Quotation
   */
  children: React.ReactNode;
  /**
   * Provide a citation for the quote
   */
  citation?: React.ReactNode;
  /**
   * Additional classes to be added to the component
   */
  className?: string;
}

export type BoxQuotationProps = RequireAtLeastOne<MinimumBoxQuotationProps, 'citation' | 'author'>;

export const BoxQuotation: FunctionComponent<BoxQuotationProps> = (props: BoxQuotationProps) => {
  const { author, children, citation, className } = props;

  const classes = classNames('ds-c-box-content-quotation', className);

  const captionContent = () => {
    // We want to prioritize citations over authors, so if both are present only render the citation.
    if (citation) {
      return <cite className="ds-c-box-content-quotation--citation">{citation}</cite>;
    }
    // If citation is not present, but author is, render the author.
    if (author && !citation) {
      return `${author} `;
    }
  };

  return (
    <figure className={classes}>
      <blockquote className="ds-c-box-content-quotation--blockquote">{children}</blockquote>
      <figcaption className="ds-c-box-content-quotation--caption">
        {`\u2014`} {captionContent()}
      </figcaption>
    </figure>
  );
};
