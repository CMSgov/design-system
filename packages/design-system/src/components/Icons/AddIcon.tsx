import React from 'react';
import SvgIcon, { IconCommonProps } from './SvgIcon';

const defaultProps = {
  className: '',
  title: 'Add',
  viewBox: '3 3 18 18',
};

function AddIcon(props: IconCommonProps): React.ReactElement {
  const iconCssClasses = `ds-c-icon--add ${props.className || ''}`;

  return (
    <SvgIcon {...defaultProps} {...props} className={iconCssClasses}>
      <path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />
    </SvgIcon>
  );
}

export default AddIcon;
