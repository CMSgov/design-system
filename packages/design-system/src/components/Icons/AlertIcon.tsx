import type * as React from 'react';
import { t } from '../i18n';
import { SvgIcon, IconCommonProps } from './SvgIcon';

const defaultProps = {
  className: '',
  viewBox: '0 0 5 17',
};

export function AlertIcon(props: IconCommonProps): React.ReactElement {
  const iconCssClasses = `ds-c-icon--alert ${props.className || ''}`;
  return (
    <SvgIcon title={t('icons.warning')} {...defaultProps} {...props} className={iconCssClasses}>
      <path d="M4.25008 15.9846C4.25008 16.1305 4.20291 16.2527 4.10932 16.3517C4.01573 16.4506 3.90104 16.4999 3.76545 16.4999H0.765403C0.630004 16.4999 0.510141 16.4479 0.406003 16.3438C0.301865 16.2397 0.249892 16.1198 0.249892 15.9846V13.016C0.249892 12.8806 0.301865 12.7607 0.406003 12.6566C0.510141 12.5525 0.630004 12.5005 0.765403 12.5005H3.76545C3.90104 12.5005 4.01592 12.55 4.10932 12.6487C4.20291 12.7481 4.25008 12.87 4.25008 13.016V15.9846ZM4.21863 10.6097C4.20808 10.7138 4.15362 10.8051 4.05485 10.8832C3.9557 10.9612 3.83353 11.0002 3.68778 11.0002H0.797047C0.6511 11.0002 0.526058 10.9612 0.421921 10.8832C0.317783 10.8051 0.265618 10.7138 0.265618 10.6097L0 0.906485C0 0.781252 0.051973 0.687662 0.156111 0.625141C0.26044 0.541908 0.385482 0.500099 0.531237 0.500099H3.96912C4.11507 0.500099 4.23992 0.541716 4.34406 0.625141C4.44819 0.687662 4.49998 0.781444 4.49998 0.906485L4.21863 10.6097Z" />
    </SvgIcon>
  );
}
