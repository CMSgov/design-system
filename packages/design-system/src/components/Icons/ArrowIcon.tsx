import React from 'react';
import SvgIcon, { SvgIconProps } from './SvgIcon';
import classNames from 'classnames';

export type ArrowIconDirectionType = 'up' | 'down' | 'left' | 'right';

export interface ArrowIconProps extends SvgIconProps {
  direction?: ArrowIconDirectionType;
}

// TODO: standardize size with other icons
function ArrowIcon(props: ArrowIconProps): React.ReactElement {
  const iconCssClasses = classNames(
    'ds-c-icon--arrow',
    `ds-c-icon--arrow-${props.direction}`,
    props.className
  );

  return (
    <SvgIcon title={`${props.direction} arrow`} {...props} className={iconCssClasses}>
      <path d="M14.502 20.77L25.37 9.894c.42-.404.631-.9.631-1.503a1.99 1.99 0 00-.631-1.502L24.118 5.62c-.41-.41-.912-.62-1.502-.62-.58 0-1.088.205-1.52.62L13 13.727 4.893 5.62C4.483 5.21 3.98 5 3.39 5c-.578 0-1.087.205-1.52.62L.62 6.888C.21 7.298 0 7.8 0 8.39s.205 1.093.62 1.503L11.48 20.77c.433.409.941.62 1.52.62.59 0 1.093-.205 1.502-.62z" />
    </SvgIcon>
  );
}

ArrowIcon.defaultProps = {
  className: '',
  direction: 'up',
  viewBox: '0 0 26 26',
};

export default ArrowIcon;
