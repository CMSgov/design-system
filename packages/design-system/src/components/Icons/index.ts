export { default as SvgIcon } from './SvgIcon';
export { default as AddIcon } from './AddIcon';
export { default as AlertCircleIcon } from './AlertCircleIcon';
export { default as ArrowsStackedIcon } from './ArrowsStacked';
export { default as ArrowIcon } from './ArrowIcon';
export { default as BuildingCircleIcon } from './BuildingCircleIcon';
export { default as CheckCircleIcon } from './CheckCircleIcon';
export { default as CheckIcon } from './CheckIcon';
export { default as CloseIcon } from './CloseIcon';
export { default as CloseIconThin } from './CloseIconThin';
export { default as DownloadIcon } from './DownloadIcon';
export { default as HHSLogo } from './HhsLogo';
export { default as ImageIcon } from './ImageIcon';
export { default as InfoCircleIcon } from './InfoCircleIcon';
export { default as InfoCircleIconThin } from './InfoCircleIconThin';
export { default as LockCircleIcon } from './LockCircleIcon';
export { default as LockIcon } from './LockIcon';
export { default as MenuIcon } from './MenuIcon';
export { default as MenuIconThin } from './MenuIconThin';
export { default as NextIcon } from './NextIcon';
export { default as PdfIcon } from './PdfIcon';
export { default as RemoveIcon } from './RemoveIcon';
export { default as StarIcon } from './StarIcon';
export { default as UsaFlagIcon } from './UsaFlagIcon';
export { default as UsaLogo } from './UsaLogo';
export { default as WarningIcon } from './WarningIcon';
export { default as WhiteHouseLogo } from './WhiteHouseLogo';
