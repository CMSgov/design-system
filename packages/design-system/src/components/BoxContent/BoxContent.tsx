import type * as React from 'react';
import { FunctionComponent } from 'react';
import classNames from 'classnames';
import { BoxQuotation } from './BoxQuotation';

export type BoxContentHeadingLevel = '1' | '2' | '3' | '4' | '5' | '6';

interface BoxContentProps {
  /**
   * Provide an author for the quote
   */
  author?: string;
  /**
   * Applies a border to the box content.
   */
  bordered?: boolean;
  /**
   * Content to be displayed within the Box
   */
  children?: React.ReactNode;
  /**
   * Provide a citation for the quote
   */
  citation?: string;
  /**
   * Additional classes to be added to the component
   */
  className?: string;
  /**
   * Text for the box content heading
   */
  heading?: React.ReactNode;
  /**
   * Heading type to override default `<h2>`.
   */
  headingLevel?: BoxContentHeadingLevel;
  /**
   * Use to highlight a direct quotation
   */
  quote?: boolean;
}

const BoxContent: FunctionComponent<BoxContentProps> = (props: BoxContentProps) => {
  const {
    author,
    bordered,
    children,
    citation,
    className,
    heading,
    headingLevel = '2',
    quote,
  } = props;

  const classes = classNames(
    'ds-c-box-content',
    bordered && 'ds-c-box-content--bordered',
    className
  );
  let headingElement;
  if (heading) {
    const Heading = `h${headingLevel}` as const;
    headingElement = <Heading className="ds-c-box-content__heading">{heading}</Heading>;
  }
  return (
    <aside className={classes}>
      <div className="ds-c-box-content__body">
        {quote ? (
          <span className="ds-text-body--lg">
            <strong>{'"'}</strong>
          </span>
        ) : (
          headingElement
        )}
        <div className={heading ? 'ds-c-box-content__text' : ''}>
          {quote ? (
            <BoxQuotation citation={citation} author={author}>
              {children}
            </BoxQuotation>
          ) : (
            children
          )}
        </div>
      </div>
    </aside>
  );
};

export default BoxContent;
