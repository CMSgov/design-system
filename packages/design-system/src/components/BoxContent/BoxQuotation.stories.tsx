import type { Meta, StoryObj } from '@storybook/react';
import BoxContent from './BoxContent';
import { BoxQuotation } from './BoxQuotation';
import { QuotationMarkIcon } from '../Icons';

const boxContentMeta: Meta = {
  title: 'Components/BoxQuotation',
  component: BoxQuotation,
  argTypes: {
    author: { control: 'text' },
    citation: { control: 'text' },
    children: { control: 'text' },
  },
};
export default boxContentMeta;
type Story = StoryObj<typeof BoxQuotation>;

const BoxQuotationStory: Story = {
  render: ({ ...args }) => {
    return (
      <BoxContent heading={args.heading}>
        <BoxQuotation {...args}>{args.children}</BoxQuotation>
      </BoxContent>
    );
  },
};

export const Default = {
  ...BoxQuotationStory,
  args: {
    citation: <a href="https://home.treasury.gov/">Inflation Reduction Act</a>,
    children:
      "The Inflation Reduction Act keeps these savings and lower costs through 2025. If you qualify for savings, you'll find out the lower costs when you shop for plans.",
    heading: <QuotationMarkIcon />,
  },
};
