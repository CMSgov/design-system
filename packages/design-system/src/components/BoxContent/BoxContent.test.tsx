import { QuotationMarkIcon } from '../Icons';
import BoxContent from './BoxContent';
import { BoxQuotation } from './BoxQuotation';
import { render, screen } from '@testing-library/react';

const defaultProps = {
  children: 'This is foo text. Bar!',
  heading: 'Test Heading',
};

function renderBoxContent(props = {}) {
  return render(<BoxContent {...defaultProps} {...props} />);
}

describe('BoxContent', () => {
  it('renders as complementary content', () => {
    renderBoxContent();
    expect(screen.getByRole('complementary')).toMatchSnapshot();
  });
  // Test that heading renders properly
  it('renders heading properly', () => {
    renderBoxContent();
    const headingElement = screen.getByText(/Test Heading/);
    expect(headingElement).toBeInTheDocument();
  });

  // Test that border renders properly
  it('renders border properly', () => {
    const { container } = renderBoxContent({ bordered: true });
    expect(container.firstChild).toHaveClass('ds-c-box-content--bordered');
  });

  // Test that quote option renders properly
  it('renders quote option properly', () => {
    renderBoxContent({
      heading: <QuotationMarkIcon />,
      children: (
        <BoxQuotation author="Test Author" citation="Test Citation">
          This is foo text. Bar!
        </BoxQuotation>
      ),
    });
    const quoteElement = screen.getByText(/This is foo text. Bar!/);
    expect(quoteElement).toBeInTheDocument();
    expect(screen.getByRole('complementary')).toMatchSnapshot();
  });

  // Test that only author renders
  it('renders a quote with only author', () => {
    renderBoxContent({
      heading: <QuotationMarkIcon />,
      children: <BoxQuotation author="Test Author">This is foo text. Bar!</BoxQuotation>,
    });
    const authorElement = screen.getByText(/Test Author/);
    expect(authorElement).toBeInTheDocument();
  });

  // Test that only citation renders
  it('renders a quote with only citation', () => {
    renderBoxContent({
      heading: <QuotationMarkIcon />,
      children: <BoxQuotation citation="Test Citation">This is foo text. Bar!</BoxQuotation>,
    });
    const citationElement = screen.getByText(/Test Citation/);
    expect(citationElement).toBeInTheDocument();
  });

  // Test that children render properly
  it('renders children properly', () => {
    renderBoxContent();
    const childrenElement = screen.getByText(/This is foo text. Bar!/);
    expect(childrenElement).toBeInTheDocument();
  });

  // Test that className is applied properly
  it('applies className properly', () => {
    const { container } = renderBoxContent({ className: 'test-class' });
    expect(container.firstChild).toHaveClass('test-class');
  });

  // Test that heading level is applied properly
  it('applies heading level properly', () => {
    renderBoxContent({ heading: 'Test Heading', headingLevel: '3' });
    const headingElement = screen.getByRole('heading', { level: 3 });
    expect(headingElement).toBeInTheDocument();
  });
});
