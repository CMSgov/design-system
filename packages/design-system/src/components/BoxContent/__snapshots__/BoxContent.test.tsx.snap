// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoxContent renders as complementary content 1`] = `
<aside
  class="ds-c-box-content"
>
  <div
    class="ds-c-box-content__body"
  >
    <h2
      class="ds-c-box-content__heading"
    >
      Test Heading
    </h2>
    <div
      class="ds-c-box-content__text"
    >
      This is foo text. Bar!
    </div>
  </div>
</aside>
`;

exports[`BoxContent renders quote option properly 1`] = `
<aside
  class="ds-c-box-content"
>
  <div
    class="ds-c-box-content__body"
  >
    <svg
      aria-hidden="true"
      class="ds-c-icon ds-c-icon--left-quote "
      viewBox="2 0 40 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M3 12.5C3 6.69875 7.67478 2 13.4464 2H14.1429C15.6837 2 16.9286 3.25125 16.9286 4.8C16.9286 6.34875 15.6837 7.6 14.1429 7.6H13.4464C10.7565 7.6 8.57143 9.79625 8.57143 12.5V13.2H14.1429C17.2158 13.2 19.7143 15.7113 19.7143 18.8V24.4C19.7143 27.4887 17.2158 30 14.1429 30H8.57143C5.49844 30 3 27.4887 3 24.4V12.5ZM25.2857 12.5C25.2857 6.69875 29.9605 2 35.7321 2H36.4286C37.9694 2 39.2143 3.25125 39.2143 4.8C39.2143 6.34875 37.9694 7.6 36.4286 7.6H35.7321C33.0422 7.6 30.8571 9.79625 30.8571 12.5V13.2H36.4286C39.5016 13.2 42 15.7113 42 18.8V24.4C42 27.4887 39.5016 30 36.4286 30H30.8571C27.7842 30 25.2857 27.4887 25.2857 24.4V12.5Z"
      />
    </svg>
    <div
      class="ds-c-box-content__text"
    >
      <figure
        class="ds-c-box-content-quotation"
      >
        <blockquote
          class="ds-c-box-content-quotation--blockquote"
        >
          This is foo text. Bar!
        </blockquote>
        <figcaption
          class="ds-c-box-content-quotation--caption"
        >
          <cite
            class="ds-c-box-content-quotation--citation"
          >
            â€”
             
            Test Citation
          </cite>
        </figcaption>
      </figure>
    </div>
  </div>
</aside>
`;
